community/cypher/cypher-compiler-2.2/src/main/scala/org/neo4j/cypher/internal/compiler/v2_2/commands/expressions/MathFunction.scala

```diff
@@ -241,15 +241,29 @@ case class RandFunction() extends Expression {
 
 case class RangeFunction(start: Expression, end: Expression, step: Expression) extends Expression with NumericHelper {
   def apply(ctx: ExecutionContext)(implicit state: QueryState): Any = {
-    val startVal = asLong(start(ctx))
-    val inclusiveEndVal = asLong(end(ctx))
     val stepVal = asLong(step(ctx))
-
     if (stepVal == 0L)
       throw new InvalidArgumentException("step argument to range() cannot be zero")
 
-    val exclusiveEndVal = inclusiveEndVal + stepVal.signum
-    (startVal until exclusiveEndVal by stepVal).toIterable
+    val startVal = asLong(start(ctx))
+    val inclusiveEndVal = asLong(end(ctx))
+    val check: (Long, Long) => Boolean = if (stepVal.signum > 0) _ <= _ else _ >= _
+
+    // due to the limitations of the scala collection library we need to implement iterator on long ranges manually:
+    // the scala one cannot be longer than MaxInt in length since it is an IndexedSeq which is indexed by Ints... :(
+    new Iterable[Long] {
```

Comment:
Interestingly you sometimes also use `range` to be later used in indexed access. So I wonder if a `toList` (for indexed access) and `toSet` (for existence check with `IN`) respectively would make sense to add as functions to cypher?